proposals/0016-simd-program-rebatable-account-write-fees.md

```diff
@@ -0,0 +1,566 @@
+of associated token account. A associated token account is a `PDA` derived from
+associated token program, where the owner is token program. But the program
+internally saves who has the authority over the account. Only token program can
+change the data of a token account. For operations like withdrawing tokens, the
+authority of the associated token account has to be the signer for the
+transaction containing transfer instruction. To receive tokens from another
+account, the authority's signature is not required.
+
+`Other Fees`: Fees other than PRAW fees `base fees + prioritization fees`.
+
+## Detailed Design
+
+PRAW fees enable dapp developer to decide who to rebate and who to penalize
+through runtime features. In particular, programs needs to be able to penalize,
+even if the application is not CPI'd to.
+
+The checks on the PRAW fees will be taken care of by Solana runtime. Total PRAW
+fees paid will be included in the transaction through a special
+`SetProgramRebatableAccountWriteFees` similar to the `ComputeBudget` program so
+that it is easier to calculate the total amount of fees that will be paid. In
+case of nonpayment or partial payment, the program is never executed. This
+instruction cannot be CPI'ed into.
+
+Addition of new syscalls to get, set and rebate PRAW fees during program
+runtime. Application developers will be able to rebate these fees through the
+new syscall, but only if the transaction succeeds, this way transactions
+write-locking the respective accounts will require full fee payment before
+program is executed.
+
+Once PRAW fees are paid on an account by the payer they are valid for the whole
+transaction, even if the same account is write-locked in multiple instructions.
+
+### Payment
+
+`SetProgramRebatableAccountWriteFees` will be part of the updated compute budget
+program. `SetProgramRebatableAccountWriteFees` will take a u64 parameter to set
+millilamports per requested CU as the PRAW fees. Instead of having separate
+instructions for each compute budget parameter we will have a single instruction
+that will set all the compute budget parameters. Compute budget parameters could
+be considered as transaction header that sets some parameters for the
+transaction. We create a new instruction that takes a vector of parameters. If
+similar transaction header parameters appear multiple times then we just take
+the first one and ignore the other occurrences. In case of multiple transaction
+headers in the transaction, we take the first occurrence and ignore the others.
+
+```
+enum TransactionHeaderParameter {
+  // existing compute budget parameters
+  SetComputeUnitLimit(u32),
+  SetComputeUnitPrice(u64),
+  SetLoadedAccountsDataSizeLimit(u32),
+  // Limit on maximum of PRAW fees to be paid
+  // Fees is set in milli lamports per CU
+  SetProgramRebatableAccountWriteFees(u64),
+}
+
+enum TransactionHeaderInstruction {
+  parameter: Vec<TransactionHeaderParameter>
+}
+```
```

Comment:
Thanks for incorporating TLV into the design. Probably need a plan to co-exist/transit from `ComputeBudgetInstruction` to ` TransactionHeaderInstruction`